// @courtlens/security/dist/utils/logger.js
"use strict";
// CourtLens Security Logger
// Centralized logging for security events with structured formatting
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecurityLogger = void 0;
const winston_1 = __importDefault(require("winston"));
class SecurityLogger {
    constructor(component) {
        this.component = component;
        this.logger = winston_1.default.createLogger({
            level: process.env.LOG_LEVEL || 'info',
            format: winston_1.default.format.combine(winston_1.default.format.timestamp(), winston_1.default.format.errors({ stack: true }), winston_1.default.format.json(), winston_1.default.format.printf(({ timestamp, level, message, component, ...meta }) => {
                return JSON.stringify({
                    timestamp,
                    level,
                    component,
                    message,
                    ...meta
                });
            })),
            defaultMeta: { component: this.component },
            transports: [
                new winston_1.default.transports.Console({
                    format: winston_1.default.format.combine(winston_1.default.format.colorize(), winston_1.default.format.simple())
                }),
                new winston_1.default.transports.File({
                    filename: 'logs/security-error.log',
                    level: 'error'
                }),
                new winston_1.default.transports.File({
                    filename: 'logs/security-combined.log'
                })
            ]
        });
        // Create logs directory if it doesn't exist
        this.ensureLogDirectory();
    }
    debug(message, meta = {}) {
        this.logger.debug(message, meta);
    }
    info(message, meta = {}) {
        this.logger.info(message, meta);
    }
    warn(message, meta = {}) {
        this.logger.warn(message, meta);
    }
    error(message, meta = {}) {
        this.logger.error(message, meta);
    }
    critical(message, meta = {}) {
        this.logger.error(message, { ...meta, level: 'critical' });
    }
    ensureLogDirectory() {
        const fs = require('fs');
        const path = require('path');
        const logDir = path.join(process.cwd(), 'logs');
        if (!fs.existsSync(logDir)) {
            fs.mkdirSync(logDir, { recursive: true });
        }
    }
}
exports.SecurityLogger = SecurityLogger;
