// @courtlens/security/dist/index.js
"use strict";
// CourtLens Security Framework - Main Export
// Central export point for all security services and utilities
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SUPPORTED_HASH_ALGORITHMS = exports.SUPPORTED_ALGORITHMS = exports.VERSION = exports.LegalSecurityUtils = exports.defaultSecurityConfig = exports.SecurityFramework = exports.SecurityValidator = exports.SecurityLogger = exports.ValidationSchemas = exports.SecurityMiddleware = exports.SecurityHardeningService = exports.EnterpriseRBACService = exports.EnterpriseAuditService = exports.SecurityEventService = exports.EncryptionService = exports.AuthorizationService = exports.AuthenticationService = void 0;
exports.createSecurityFramework = createSecurityFramework;
exports.getSecurityHealthStatus = getSecurityHealthStatus;
// Types
__exportStar(require("./types"), exports);
// Core Services
var authentication_service_1 = require("./services/authentication.service");
Object.defineProperty(exports, "AuthenticationService", { enumerable: true, get: function () { return authentication_service_1.AuthenticationService; } });
var authorization_service_1 = require("./services/authorization.service");
Object.defineProperty(exports, "AuthorizationService", { enumerable: true, get: function () { return authorization_service_1.AuthorizationService; } });
var encryption_service_1 = require("./services/encryption.service");
Object.defineProperty(exports, "EncryptionService", { enumerable: true, get: function () { return encryption_service_1.EncryptionService; } });
var security_event_service_1 = require("./services/security-event.service");
Object.defineProperty(exports, "SecurityEventService", { enumerable: true, get: function () { return security_event_service_1.SecurityEventService; } });
// Enterprise Services
var enterprise_audit_service_1 = require("./services/enterprise-audit.service");
Object.defineProperty(exports, "EnterpriseAuditService", { enumerable: true, get: function () { return enterprise_audit_service_1.EnterpriseAuditService; } });
var enterprise_rbac_service_1 = require("./services/enterprise-rbac.service");
Object.defineProperty(exports, "EnterpriseRBACService", { enumerable: true, get: function () { return enterprise_rbac_service_1.EnterpriseRBACService; } });
var security_hardening_service_1 = require("./services/security-hardening.service");
Object.defineProperty(exports, "SecurityHardeningService", { enumerable: true, get: function () { return security_hardening_service_1.SecurityHardeningService; } });
// Middleware
var security_middleware_1 = require("./middleware/security.middleware");
Object.defineProperty(exports, "SecurityMiddleware", { enumerable: true, get: function () { return security_middleware_1.SecurityMiddleware; } });
Object.defineProperty(exports, "ValidationSchemas", { enumerable: true, get: function () { return security_middleware_1.ValidationSchemas; } });
// Utilities
var logger_1 = require("./utils/logger");
Object.defineProperty(exports, "SecurityLogger", { enumerable: true, get: function () { return logger_1.SecurityLogger; } });
const authentication_service_2 = require("./services/authentication.service");
const authorization_service_2 = require("./services/authorization.service");
const encryption_service_2 = require("./services/encryption.service");
const security_middleware_2 = require("./middleware/security.middleware");
const security_event_service_2 = require("./services/security-event.service");
const enterprise_audit_service_2 = require("./services/enterprise-audit.service");
const enterprise_rbac_service_2 = require("./services/enterprise-rbac.service");
const security_hardening_service_2 = require("./services/security-hardening.service");
var validator_1 = require("./utils/validator");
Object.defineProperty(exports, "SecurityValidator", { enumerable: true, get: function () { return validator_1.SecurityValidator; } });
// Security Framework Factory
class SecurityFramework {
    constructor(config) {
        // Initialize core services
        this.authentication = new authentication_service_2.AuthenticationService(config.supabase.url, config.supabase.serviceRoleKey, config.jwt.secret);
        this.authorization = new authorization_service_2.AuthorizationService(config.supabase.url, config.supabase.serviceRoleKey);
        this.encryption = new encryption_service_2.EncryptionService();
        this.eventService = new security_event_service_2.SecurityEventService(config.supabase.url, config.supabase.serviceRoleKey);
        // Initialize enterprise services if enabled
        if (config.enterprise?.enableAdvancedAudit) {
            this.enterpriseAudit = new enterprise_audit_service_2.EnterpriseAuditService();
        }
        if (config.enterprise?.enableAdvancedRBAC && this.enterpriseAudit) {
            this.enterpriseRBAC = new enterprise_rbac_service_2.EnterpriseRBACService(this.enterpriseAudit);
        }
        if (config.enterprise?.enableSecurityHardening && this.enterpriseAudit) {
            this.securityHardening = new security_hardening_service_2.SecurityHardeningService(this.enterpriseAudit);
        }
        this.middleware = new security_middleware_2.SecurityMiddleware({
            authenticationService: this.authentication,
            authorizationService: this.authorization
        });
    }
    /**
     * Get complete security configuration for Express app
     */
    getExpressConfig() {
        return {
            // Security headers
            securityHeaders: this.middleware.securityHeaders(),
            // Rate limiting
            rateLimit: this.middleware.rateLimit(),
            authRateLimit: this.middleware.authRateLimit(),
            // CORS
            cors: this.middleware.cors(),
            // Authentication & Authorization
            authenticate: this.middleware.authenticate.bind(this.middleware),
            authorize: this.middleware.authorize.bind(this.middleware),
            // Validation
            validateInput: this.middleware.validateInput.bind(this.middleware),
            // Utilities
            requireOrganizationAccess: this.middleware.requireOrganizationAccess.bind(this.middleware),
            // Error handling
            errorHandler: this.middleware.errorHandler(),
            requestLogger: this.middleware.requestLogger()
        };
    }
}
exports.SecurityFramework = SecurityFramework;
// Default configuration
exports.defaultSecurityConfig = {
    supabase: {
        url: process.env.SUPABASE_URL || 'https://qzkopgqmymorpngpabvq.supabase.co',
        serviceRoleKey: process.env.SUPABASE_SERVICE_ROLE_KEY || '',
        anonKey: process.env.SUPABASE_ANON_KEY || ''
    },
    jwt: {
        secret: process.env.JWT_SECRET || 'your-super-secret-jwt-key',
        expiresIn: process.env.JWT_EXPIRES_IN || '1h'
    },
    encryption: {
        defaultLevel: 'enhanced'
    },
    monitoring: {
        enableRealTimeAlerts: true,
        logLevel: 'info'
    }
};
// Factory function for easy initialization
function createSecurityFramework(config) {
    const finalConfig = { ...exports.defaultSecurityConfig, ...config };
    return new SecurityFramework(finalConfig);
}
// Legal Domain Security Utilities
class LegalSecurityUtils {
    constructor(encryptionService) {
        this.encryption = encryptionService;
    }
    /**
     * Encrypt legal document content
     */
    async encryptLegalDocument(content, classification) {
        const level = this.getEncryptionLevelForClassification(classification);
        const encryptedData = await this.encryption.encrypt(content, level);
        return JSON.stringify(encryptedData);
    }
    /**
     * Generate secure document ID
     */
    generateSecureDocumentId() {
        return this.encryption.generateToken(16);
    }
    /**
     * Hash client information for privacy
     */
    hashClientInfo(clientData) {
        return this.encryption.hash(clientData);
    }
    getEncryptionLevelForClassification(classification) {
        // Map document classification to encryption level
        const levelMap = {
            'public': 'standard',
            'internal': 'standard',
            'confidential': 'enhanced',
            'restricted': 'maximum',
            'top_secret': 'maximum'
        };
        return levelMap[classification] || 'enhanced';
    }
}
exports.LegalSecurityUtils = LegalSecurityUtils;
// Export version information
exports.VERSION = '1.0.0';
exports.SUPPORTED_ALGORITHMS = ['aes-256-gcm', 'aes-256-cbc'];
exports.SUPPORTED_HASH_ALGORITHMS = ['sha256', 'sha512'];
// Health check utility
function getSecurityHealthStatus() {
    const checks = [
        {
            name: 'environment_variables',
            status: !!(process.env.SUPABASE_URL && process.env.JWT_SECRET),
            message: 'Required environment variables present'
        },
        {
            name: 'encryption_service',
            status: true, // Could add actual health check
            message: 'Encryption service operational'
        },
        {
            name: 'authentication_service',
            status: true, // Could add actual health check
            message: 'Authentication service operational'
        }
    ];
    const allHealthy = checks.every(check => check.status);
    const anyUnhealthy = checks.some(check => !check.status);
    return {
        status: allHealthy ? 'healthy' : anyUnhealthy ? 'unhealthy' : 'degraded',
        checks
    };
}
