// @courtlens/security/src/index.ts
// CourtLens Security Framework - Main Export
// Central export point for all security services and utilities

// Types
export * from './types';

// Core Services
export { AuthenticationService } from './services/authentication.service';
export { AuthorizationService } from './services/authorization.service';
export { EncryptionService } from './services/encryption.service';
export { SecurityEventService } from './services/security-event.service';

// Enterprise Services
export { EnterpriseAuditService } from './services/enterprise-audit.service';
export { EnterpriseRBACService } from './services/enterprise-rbac.service';
export { SecurityHardeningService } from './services/security-hardening.service';

// Middleware
export { SecurityMiddleware, ValidationSchemas } from './middleware/security.middleware';

// Utilities
export { SecurityLogger } from './utils/logger';
import { SecurityConfig, DocumentClassification, EncryptionLevel } from './types/index';
import { createClient } from '@supabase/supabase-js';
import { SecurityLogger } from './utils/logger';
import winston from 'winston';
import { AuthenticationService } from './services/authentication.service';
import { AuthorizationService } from './services/authorization.service';
import { EncryptionService } from './services/encryption.service';
import { SecurityMiddleware } from './middleware/security.middleware';
import { SecurityEventService } from './services/security-event.service';
import { EnterpriseAuditService } from './services/enterprise-audit.service';
import { EnterpriseRBACService } from './services/enterprise-rbac.service';
import { SecurityHardeningService } from './services/security-hardening.service';
export { SecurityValidator } from './utils/validator';

// Security Configuration
export interface SecurityFrameworkConfig {
  supabase: {
    url: string;
    serviceRoleKey: string;
    anonKey: string;
  };
  jwt: {
    secret: string;
    expiresIn: string;
  };
  encryption: {
    defaultLevel: 'standard' | 'enhanced' | 'maximum';
  };
  monitoring: {
    enableRealTimeAlerts: boolean;
    logLevel: 'debug' | 'info' | 'warn' | 'error' | 'critical';
  };
  enterprise?: {
    enableAdvancedAudit: boolean;
    enableAdvancedRBAC: boolean;
    enableSecurityHardening: boolean;
  };
}

// Security Framework Factory
export class SecurityFramework {
  public readonly authentication: AuthenticationService;
  public readonly authorization: AuthorizationService;
  public readonly encryption: EncryptionService;
  public readonly middleware: SecurityMiddleware;
  public readonly eventService: SecurityEventService;
  
  // Enterprise Services
  public readonly enterpriseAudit?: EnterpriseAuditService;
  public readonly enterpriseRBAC?: EnterpriseRBACService;
  public readonly securityHardening?: SecurityHardeningService;

  constructor(config: SecurityFrameworkConfig) {
    // Initialize core services
    this.authentication = new AuthenticationService(
      config.supabase.url,
      config.supabase.serviceRoleKey,
      config.jwt.secret
    );

    this.authorization = new AuthorizationService(
      config.supabase.url,
      config.supabase.serviceRoleKey
    );

    this.encryption = new EncryptionService();

    this.eventService = new SecurityEventService(
      config.supabase.url,
      config.supabase.serviceRoleKey
    );

    // Initialize enterprise services if enabled
    if (config.enterprise?.enableAdvancedAudit) {
      this.enterpriseAudit = new EnterpriseAuditService();
    }

    if (config.enterprise?.enableAdvancedRBAC && this.enterpriseAudit) {
      this.enterpriseRBAC = new EnterpriseRBACService(this.enterpriseAudit);
    }

    if (config.enterprise?.enableSecurityHardening && this.enterpriseAudit) {
      this.securityHardening = new SecurityHardeningService(this.enterpriseAudit);
    }

    this.middleware = new SecurityMiddleware({
      authenticationService: this.authentication,
      authorizationService: this.authorization
    });
  }

  /**
   * Get complete security configuration for Express app
   */
  getExpressConfig() {
    return {
      // Security headers
      securityHeaders: this.middleware.securityHeaders(),
      
      // Rate limiting
      rateLimit: this.middleware.rateLimit(),
      authRateLimit: this.middleware.authRateLimit(),
      
      // CORS
      cors: this.middleware.cors(),
      
      // Authentication & Authorization
      authenticate: this.middleware.authenticate.bind(this.middleware),
      authorize: this.middleware.authorize.bind(this.middleware),
      
      // Validation
      validateInput: this.middleware.validateInput.bind(this.middleware),
      
      // Utilities
      requireOrganizationAccess: this.middleware.requireOrganizationAccess.bind(this.middleware),
      
      // Error handling
      errorHandler: this.middleware.errorHandler(),
      requestLogger: this.middleware.requestLogger()
    };
  }
}

// Default configuration
export const defaultSecurityConfig: SecurityFrameworkConfig = {
  supabase: {
    url: process.env.SUPABASE_URL || 'https://qzkopgqmymorpngpabvq.supabase.co',
    serviceRoleKey: process.env.SUPABASE_SERVICE_ROLE_KEY || '',
    anonKey: process.env.SUPABASE_ANON_KEY || ''
  },
  jwt: {
    secret: process.env.JWT_SECRET || 'your-super-secret-jwt-key',
    expiresIn: process.env.JWT_EXPIRES_IN || '1h'
  },
  encryption: {
    defaultLevel: 'enhanced'
  },
  monitoring: {
    enableRealTimeAlerts: true,
    logLevel: 'info'
  }
};

// Factory function for easy initialization
export function createSecurityFramework(config?: Partial<SecurityFrameworkConfig>): SecurityFramework {
  const finalConfig = { ...defaultSecurityConfig, ...config };
  return new SecurityFramework(finalConfig);
}

// Legal Domain Security Utilities
export class LegalSecurityUtils {
  private encryption: EncryptionService;

  constructor(encryptionService: EncryptionService) {
    this.encryption = encryptionService;
  }

  /**
   * Encrypt legal document content
   */
  async encryptLegalDocument(content: string, classification: DocumentClassification): Promise<string> {
    const level = this.getEncryptionLevelForClassification(classification);
    const encryptedData = await this.encryption.encrypt(content, level);
    return JSON.stringify(encryptedData);
  }

  /**
   * Generate secure document ID
   */
  generateSecureDocumentId(): string {
    return this.encryption.generateToken(16);
  }

  /**
   * Hash client information for privacy
   */
  hashClientInfo(clientData: string): string {
    return this.encryption.hash(clientData);
  }

  private getEncryptionLevelForClassification(classification: string): any {
    // Map document classification to encryption level
    const levelMap: Record<string, any> = {
      'public': 'standard',
      'internal': 'standard',
      'confidential': 'enhanced',
      'restricted': 'maximum',
      'top_secret': 'maximum'
    };
    
    return levelMap[classification] || 'enhanced';
  }
}

// Export version information
export const VERSION = '1.0.0';
export const SUPPORTED_ALGORITHMS = ['aes-256-gcm', 'aes-256-cbc'];
export const SUPPORTED_HASH_ALGORITHMS = ['sha256', 'sha512'];

// Health check utility
export function getSecurityHealthStatus(): {
  status: 'healthy' | 'degraded' | 'unhealthy';
  checks: Array<{ name: string; status: boolean; message?: string }>;
} {
  const checks = [
    {
      name: 'environment_variables',
      status: !!(process.env.SUPABASE_URL && process.env.JWT_SECRET),
      message: 'Required environment variables present'
    },
    {
      name: 'encryption_service',
      status: true, // Could add actual health check
      message: 'Encryption service operational'
    },
    {
      name: 'authentication_service',
      status: true, // Could add actual health check
      message: 'Authentication service operational'
    }
  ];

  const allHealthy = checks.every(check => check.status);
  const anyUnhealthy = checks.some(check => !check.status);

  return {
    status: allHealthy ? 'healthy' : anyUnhealthy ? 'unhealthy' : 'degraded',
    checks
  };
}
