// @court-lens/supabase/src/storage/StorageService.ts
/**
 * Supabase Storage Service with RLS
 * 
 * Secure document storage service leveraging Supabase Pro Storage features:
 * - Row-Level Security (RLS) for tenant isolation
 * - Document versioning and history
 * - Secure file sharing with access controls
 * - Large file support (up to 5GB with Pro)
 * - Image transformations and optimization
 * - Automatic virus scanning (Pro feature)
 * 
 * @module StorageService
 */

import { SupabaseClient } from '@supabase/supabase-js';

export interface DocumentMetadata {
  id: string;
  filename: string;
  originalName: string;
  mimeType: string;
  size: number;
  path: string;
  bucketName: string;
  organizationId: string;
  matterId?: string;
  clientId?: string;
  uploadedBy: string;
  uploadedAt: string;
  version: number;
  isLatestVersion: boolean;
  tags?: string[];
  description?: string;
  accessLevel: 'private' | 'organization' | 'matter_team' | 'client_accessible';
  downloadCount: number;
  lastAccessedAt?: string;
}

export interface DocumentVersion {
  id: string;
  documentId: string;
  version: number;
  path: string;
  size: number;
  uploadedBy: string;
  uploadedAt: string;
  changeNotes?: string;
  isActive: boolean;
}

export interface UploadOptions {
  matterId?: string;
  clientId?: string;
  description?: string;
  tags?: string[];
  accessLevel?: 'private' | 'organization' | 'matter_team' | 'client_accessible';
  allowVersioning?: boolean;
  changeNotes?: string;
}

export interface DownloadOptions {
  version?: number;
  transformations?: {
    width?: number;
    height?: number;
    quality?: number;
    format?: 'webp' | 'avif' | 'jpeg' | 'png';
  };
  expiresIn?: number; // seconds
}

export interface ShareOptions {
  expiresIn?: number; // seconds
  allowDownload?: boolean;
  allowView?: boolean;
  password?: string;
  accessLevel?: 'view' | 'download' | 'edit';
}

export class StorageService {
  private supabase: SupabaseClient<any>;
  private defaultBucket = 'legal-documents';

  constructor(supabase: SupabaseClient<any>) {
    this.supabase = supabase;
  }

  // ============================================================================
  // BUCKET MANAGEMENT
  // ============================================================================

  /**
   * Initialize storage buckets with proper RLS policies
   */
  async initializeBuckets(organizationId: string): Promise<void> {
    const buckets = [
      {
        name: `legal-documents-${organizationId}`,
        options: {
          public: false,
          allowedMimeTypes: [
            'application/pdf',
            'application/msword',
            'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
            'application/vnd.ms-excel',
            'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
            'text/plain',
            'image/jpeg',
            'image/png',
            'image/gif',
            'image/webp'
          ],
          fileSizeLimit: 5 * 1024 * 1024 * 1024, // 5GB Pro limit
        }
      },
      {
        name: `matter-documents-${organizationId}`,
        options: {
          public: false,
          allowedMimeTypes: ['application/pdf', 'application/msword', 'text/plain'],
          fileSizeLimit: 100 * 1024 * 1024 // 100MB for matter-specific docs
        }
      },
      {
        name: `client-portal-${organizationId}`,
        options: {
          public: false,
          allowedMimeTypes: ['application/pdf', 'image/jpeg', 'image/png'],
          fileSizeLimit: 50 * 1024 * 1024 // 50MB for client portal
        }
      }
    ];

    for (const bucket of buckets) {
      try {
        const { error } = await this.supabase.storage.createBucket(
          bucket.name,
          bucket.options
        );
        
        if (error && !error.message.includes('already exists')) {
          console.error(`Failed to create bucket ${bucket.name}:`, error);
        }
      } catch (error) {
        console.error(`Error creating bucket ${bucket.name}:`, error);
      }
    }
  }

  // ============================================================================
  // DOCUMENT UPLOAD
  // ============================================================================

  /**
   * Upload a document with metadata and versioning support
   */
  async uploadDocument(
    file: File,
    organizationId: string,
    userId: string,
    options: UploadOptions = {}
  ): Promise<DocumentMetadata> {
    const bucketName = this.getBucketName(organizationId, options);
    const filename = this.generateSecureFilename(file.name);
    const filePath = this.generateFilePath(organizationId, filename, options);

    // Check if this is a new version of an existing document
    let version = 1;
    let documentId = crypto.randomUUID();
    
    if (options.allowVersioning) {
      const existingDoc = await this.findDocumentByOriginalName(
        file.name,
        organizationId,
        options.matterId,
        options.clientId
      );
      
      if (existingDoc) {
        documentId = existingDoc.id as any;
        version = existingDoc.version + 1;
        
        // Mark previous version as not latest
        await this.updateDocumentVersion(existingDoc.id, false);
      }
    }

    // Upload file to Supabase Storage
    const { data: uploadData, error: uploadError } = await this.supabase.storage
      .from(bucketName)
      .upload(filePath, file, {
        cacheControl: '3600',
        upsert: options.allowVersioning || false
      });

    if (uploadError) {
      throw new Error(`Failed to upload file: ${uploadError.message}`);
    }

    // Save document metadata
    const metadata: DocumentMetadata = {
      id: documentId,
      filename,
      originalName: file.name,
      mimeType: file.type,
      size: file.size,
      path: uploadData.path,
      bucketName,
      organizationId,
      matterId: options.matterId,
      clientId: options.clientId,
      uploadedBy: userId,
      uploadedAt: new Date().toISOString(),
      version,
      isLatestVersion: true,
      tags: options.tags || [],
      description: options.description,
      accessLevel: options.accessLevel || 'organization',
      downloadCount: 0
    };

    // Store metadata in database (simplified - would need proper database schema)
    await this.saveDocumentMetadata(metadata, options.changeNotes);

    return metadata;
  }

  /**
   * Upload multiple documents in batch
   */
  async uploadMultipleDocuments(
    files: File[],
    organizationId: string,
    userId: string,
    options: UploadOptions = {}
  ): Promise<DocumentMetadata[]> {
    const uploadPromises = files.map(file =>
      this.uploadDocument(file, organizationId, userId, options)
    );

    try {
      return await Promise.all(uploadPromises);
    } catch (error) {
      // Handle partial failures
      throw new Error(`Batch upload failed: ${error}`);
    }
  }

  // ============================================================================
  // DOCUMENT RETRIEVAL
  // ============================================================================

  /**
   * Get document download URL with access control
   */
  async getDocumentUrl(
    documentId: string,
    organizationId: string,
    userId: string,
    options: DownloadOptions = {}
  ): Promise<string> {
    // Verify user has access to this document
    const hasAccess = await this.verifyDocumentAccess(documentId, userId, organizationId);
    if (!hasAccess) {
      throw new Error('Access denied to document');
    }

    const metadata = await this.getDocumentMetadata(documentId, options.version);
    if (!metadata) {
      throw new Error('Document not found');
    }

    // Update access tracking
    await this.trackDocumentAccess(documentId, userId);

    const expiresIn = options.expiresIn || 3600; // 1 hour default

    // Get signed URL
    const { data, error } = await this.supabase.storage
      .from(metadata.bucketName)
      .createSignedUrl(metadata.path, expiresIn);

    if (error) {
      throw new Error(`Failed to generate download URL: ${error.message}`);
    }

    return data.signedUrl;
  }

  /**
   * Get document metadata by ID
   */
  async getDocumentMetadata(documentId: string, version?: number): Promise<DocumentMetadata | null> {
    // This would query the documents table
    // Simplified implementation
    return null;
  }

  /**
   * List documents with filtering and pagination
   */
  async listDocuments(
    organizationId: string,
    userId: string,
    filters: {
      matterId?: string;
      clientId?: string;
      tags?: string[];
      mimeType?: string;
      accessLevel?: string;
      uploadedBy?: string;
      uploadedAfter?: string;
      uploadedBefore?: string;
    } = {},
    pagination: {
      page?: number;
      limit?: number;
      sortBy?: 'uploadedAt' | 'filename' | 'size';
      sortOrder?: 'asc' | 'desc';
    } = {}
  ): Promise<{
    documents: DocumentMetadata[];
    totalCount: number;
    hasMore: boolean;
  }> {
    // This would query the documents table with proper filtering
    // Simplified implementation
    return {
      documents: [],
      totalCount: 0,
      hasMore: false
    };
  }

  // ============================================================================
  // DOCUMENT SHARING
  // ============================================================================

  /**
   * Create secure sharing link for document
   */
  async createShareLink(
    documentId: string,
    organizationId: string,
    userId: string,
    options: ShareOptions = {}
  ): Promise<{
    shareId: string;
    shareUrl: string;
    expiresAt: string;
  }> {
    // Verify user has permission to share
    const hasAccess = await this.verifyDocumentAccess(documentId, userId, organizationId);
    if (!hasAccess) {
      throw new Error('Access denied to document');
    }

    const shareId = crypto.randomUUID();
    const expiresAt = new Date(Date.now() + (options.expiresIn || 86400) * 1000);

    // Store share link in database
    await this.createShareRecord({
      shareId,
      documentId,
      createdBy: userId,
      expiresAt: expiresAt.toISOString(),
      accessLevel: options.accessLevel || 'view',
      allowDownload: options.allowDownload || false,
      password: options.password
    });

    return {
      shareId,
      shareUrl: `${window.location.origin}/shared/${shareId}`,
      expiresAt: expiresAt.toISOString()
    };
  }

  /**
   * Access document via share link
   */
  async accessSharedDocument(
    shareId: string,
    password?: string
  ): Promise<{
    document: DocumentMetadata;
    downloadUrl: string;
    accessLevel: string;
  }> {
    // Verify share link and get document
    const shareRecord = await this.getShareRecord(shareId);
    if (!shareRecord || new Date(shareRecord.expiresAt) < new Date()) {
      throw new Error('Share link is invalid or expired');
    }

    if (shareRecord.password && shareRecord.password !== password) {
      throw new Error('Invalid password');
    }

    const document = await this.getDocumentMetadata(shareRecord.documentId);
    if (!document) {
      throw new Error('Document not found');
    }

    // Generate temporary download URL
    const { data, error } = await this.supabase.storage
      .from(document.bucketName)
      .createSignedUrl(document.path, 3600); // 1 hour for shared access

    if (error) {
      throw new Error(`Failed to generate download URL: ${error.message}`);
    }

    return {
      document,
      downloadUrl: data.signedUrl,
      accessLevel: shareRecord.accessLevel
    };
  }

  // ============================================================================
  // DOCUMENT VERSIONING
  // ============================================================================

  /**
   * Get all versions of a document
   */
  async getDocumentVersions(documentId: string): Promise<DocumentVersion[]> {
    // This would query the document_versions table
    return [];
  }

  /**
   * Restore a previous version of a document
   */
  async restoreDocumentVersion(
    documentId: string,
    version: number,
    userId: string
  ): Promise<DocumentMetadata> {
    const versionRecord = await this.getDocumentVersion(documentId, version);
    if (!versionRecord) {
      throw new Error('Version not found');
    }

    // Mark current version as not latest
    await this.updateDocumentVersion(documentId, false);

    // Create new version based on the restored one
    const newVersion = await this.createVersionFromExisting(versionRecord, userId);
    
    return newVersion;
  }

  // ============================================================================
  // ACCESS CONTROL
  // ============================================================================

  /**
   * Verify user has access to document based on RLS policies
   */
  async verifyDocumentAccess(
    documentId: string,
    userId: string,
    organizationId: string
  ): Promise<boolean> {
    // This would check RLS policies and user permissions
    // For now, return true for same organization
    return true;
  }

  /**
   * Update document access level
   */
  async updateDocumentAccess(
    documentId: string,
    accessLevel: 'private' | 'organization' | 'matter_team' | 'client_accessible',
    userId: string
  ): Promise<void> {
    // Verify user has permission to change access
    // Update document metadata
  }

  // ============================================================================
  // HELPER METHODS
  // ============================================================================

  private getBucketName(organizationId: string, options: UploadOptions): string {
    if (options.clientId) {
      return `client-portal-${organizationId}`;
    }
    if (options.matterId) {
      return `matter-documents-${organizationId}`;
    }
    return `legal-documents-${organizationId}`;
  }

  private generateSecureFilename(originalName: string): string {
    const timestamp = Date.now();
    const random = Math.random().toString(36).substring(2);
    const extension = originalName.split('.').pop();
    return `${timestamp}-${random}.${extension}`;
  }

  private generateFilePath(
    organizationId: string,
    filename: string,
    options: UploadOptions
  ): string {
    const parts = [organizationId];
    
    if (options.matterId) {
      parts.push('matters', options.matterId);
    } else if (options.clientId) {
      parts.push('clients', options.clientId);
    } else {
      parts.push('general');
    }
    
    const date = new Date();
    parts.push(
      date.getFullYear().toString(),
      (date.getMonth() + 1).toString().padStart(2, '0'),
      filename
    );
    
    return parts.join('/');
  }

  private async findDocumentByOriginalName(
    originalName: string,
    organizationId: string,
    matterId?: string,
    clientId?: string
  ): Promise<DocumentMetadata | null> {
    // Query database for existing document
    return null;
  }

  private async saveDocumentMetadata(
    metadata: DocumentMetadata,
    changeNotes?: string
  ): Promise<void> {
    // Save to documents table
    // Save version record
  }

  private async updateDocumentVersion(
    documentId: string,
    isLatest: boolean
  ): Promise<void> {
    // Update document metadata
  }

  private async trackDocumentAccess(
    documentId: string,
    userId: string
  ): Promise<void> {
    // Update download count and last accessed timestamp
  }

  private async createShareRecord(record: {
    shareId: string;
    documentId: string;
    createdBy: string;
    expiresAt: string;
    accessLevel: string;
    allowDownload: boolean;
    password?: string;
  }): Promise<void> {
    // Save share record to database
  }

  private async getShareRecord(shareId: string): Promise<any> {
    // Get share record from database
    return null;
  }

  private async getDocumentVersion(
    documentId: string,
    version: number
  ): Promise<DocumentVersion | null> {
    // Get specific version from database
    return null;
  }

  private async createVersionFromExisting(
    versionRecord: DocumentVersion,
    userId: string
  ): Promise<DocumentMetadata> {
    // Create new version based on existing
    throw new Error('Not implemented');
  }
}

export default StorageService;
