// @courtlens/security/src/utils/logger.ts
// CourtLens Security Logger
// Centralized logging for security events with structured formatting

import winston from 'winston';
import { LogLevel } from '../types';

export class SecurityLogger {
  private logger: winston.Logger;
  private component: string;

  constructor(component: string) {
    this.component = component;
    this.logger = winston.createLogger({
      level: process.env.LOG_LEVEL || 'info',
      format: winston.format.combine(
        winston.format.timestamp(),
        winston.format.errors({ stack: true }),
        winston.format.json(),
        winston.format.printf(({ timestamp, level, message, component, ...meta }) => {
          return JSON.stringify({
            timestamp,
            level,
            component,
            message,
            ...meta
          });
        })
      ),
      defaultMeta: { component: this.component },
      transports: [
        new winston.transports.Console({
          format: winston.format.combine(
            winston.format.colorize(),
            winston.format.simple()
          )
        }),
        new winston.transports.File({
          filename: 'logs/security-error.log',
          level: 'error'
        }),
        new winston.transports.File({
          filename: 'logs/security-combined.log'
        })
      ]
    });

    // Create logs directory if it doesn't exist
    this.ensureLogDirectory();
  }

  debug(message: string, meta: Record<string, any> = {}): void {
    this.logger.debug(message, meta);
  }

  info(message: string, meta: Record<string, any> = {}): void {
    this.logger.info(message, meta);
  }

  warn(message: string, meta: Record<string, any> = {}): void {
    this.logger.warn(message, meta);
  }

  error(message: string, meta: Record<string, any> = {}): void {
    this.logger.error(message, meta);
  }

  critical(message: string, meta: Record<string, any> = {}): void {
    this.logger.error(message, { ...meta, level: 'critical' });
  }

  private ensureLogDirectory(): void {
    const fs = require('fs');
    const path = require('path');
    
    const logDir = path.join(process.cwd(), 'logs');
    if (!fs.existsSync(logDir)) {
      fs.mkdirSync(logDir, { recursive: true });
    }
  }
}
