// @courtlens/security/src/utils/validator.ts
// CourtLens Security Validator
// Input validation and sanitization for security-critical operations

import Joi from 'joi';
import { ValidationError } from '../types';

export class SecurityValidator {
  private emailSchema = Joi.string().email().required();
  private passwordSchema = Joi.string()
    .min(8)
    .max(128)
    .pattern(/^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]/)
    .required()
    .messages({
      'string.pattern.base': 'Password must contain at least one lowercase letter, one uppercase letter, one number, and one special character'
    });

  validateEmail(email: string): void {
    const { error } = this.emailSchema.validate(email);
    if (error) {
      throw new ValidationError(`Invalid email: ${error.details[0].message}`);
    }
  }

  validatePassword(password: string): void {
    const { error } = this.passwordSchema.validate(password);
    if (error) {
      throw new ValidationError(`Invalid password: ${error.details[0].message}`);
    }
  }

  validateRequired(value: any, fieldName: string): void {
    if (!value || (typeof value === 'string' && value.trim().length === 0)) {
      throw new ValidationError(`${fieldName} is required`);
    }
  }

  sanitizeInput(input: string): string {
    return input
      .replace(/[<>]/g, '') // Remove potential HTML tags
      .replace(/['"]/g, '') // Remove quotes
      .trim();
  }

  validateIpAddress(ip: string): boolean {
    const ipv4Regex = /^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/;
    const ipv6Regex = /^(?:[0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}$/;
    return ipv4Regex.test(ip) || ipv6Regex.test(ip);
  }

  validateUuid(uuid: string): boolean {
    const uuidRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
    return uuidRegex.test(uuid);
  }
}
